NitroxClient/MonoBehaviours/Gui/MainMenu/ServerJoin/JoinServerBackend.cs

```diff
@@ -0,0 +1,170 @@
+
+                break;
+
+            case MultiplayerSessionConnectionStage.SESSION_RESERVATION_REJECTED:
+                Log.Info("Reservation rejected");
+                Log.InGame(Language.main.Get("Nitrox_RejectedSessionPolicy"));
+
+                MultiplayerSessionReservationState reservationState = multiplayerSession.Reservation.ReservationState;
+
+                string reservationRejectionNotification = reservationState.Describe();
+
+                MainMenuNotificationPanel.ShowMessage(reservationRejectionNotification, null, () =>
+                {
+                    multiplayerSession.Disconnect();
+                    multiplayerSession.ConnectAsync(serverIp, serverPort);
+                });
+
+                break;
+
+            case MultiplayerSessionConnectionStage.DISCONNECTED:
+                Log.Info(Language.main.Get("Nitrox_DisconnectedSession"));
+                break;
+        }
+    }
+
+    [SuppressMessage("Usage", "DIMA001:Dependency Injection container is used directly")]
+    public static async Task StartMultiplayerClientAsync(string ip, int port)
+    {
+        serverIp = ip;
+        serverPort = port;
+        NitroxServiceLocator.BeginNewLifetimeScope();
+
+        preferencesManager = NitroxServiceLocator.LocateService<PlayerPreferenceManager>();
+        activePlayerPreference = preferencesManager.GetPreference(serverIp);
+        multiplayerSession = NitroxServiceLocator.LocateService<IMultiplayerSession>();
+
+        if (!multiplayerClient)
+        {
+            multiplayerClient = new GameObject("Nitrox Multiplayer Client");
+            multiplayerClient.AddComponent<Multiplayer>();
+            multiplayerSession.ConnectionStateChanged += SessionConnectionStateChangedHandler;
+        }
+
+        try
+        {
+            await multiplayerSession.ConnectAsync(serverIp, serverPort);
+        }
+        catch (ClientConnectionFailedException ex)
+        {
+            Log.ErrorSensitive("Unable to contact the remote server at: {ip}:{port}", serverIp, serverPort);
+            string msg = $"{Language.main.Get("Nitrox_UnableToConnect")} {serverIp}:{serverPort}";
+
+            if (serverIp.Equals("127.0.0.1"))
+            {
+                if (Process.GetProcessesByName("NitroxServer-Subnautica").Length == 0)
+                {
+                    Log.Error("No server process was found while address was 127.0.0.1");
+                    msg += $"\n{Language.main.Get("Nitrox_StartServer")}";
+                }
+                else
+                {
+                    Log.Error(ex);
+                    msg += $"\n{Language.main.Get("Nitrox_FirewallInterfering")}";
+                }
+            }
+
+            Log.InGame(msg);
+            StopMultiplayerClient();
+            MainMenuNotificationPanel.ShowMessage(msg, MainMenuServerListPanel.NAME);
+        }
+    }
+
+    [SuppressMessage("Usage", "DIMA001:Dependency Injection container is used directly")]
```

Comment:
This shouldn't be suppressed as we can fix it with proper architecture.

Reply:
The thing is that the LifetimeScope is started in this method so constructor injection seems over-complicated imo.

Reply:
If you change the end of `MainMenuServerButton.OpenJoinServerMenuAsync` to this you can remove the suppress and we'll fix the DI usage later.
```cs
NitroxServiceLocator.BeginNewLifetimeScope();
await JoinServerBackend.StartMultiplayerClientAsync(endpoint.Address.ToString(), endpoint.Port);
```